middleware/queue/source/example/main.cpp

```diff
@@ -0,0 +1,104 @@
+#include "common/argument.h"
+#include "common/communication/ipc.h"
+#include "common/communication/instance.h"
+#include "common/exception/guard.h"
+#include "common/message/dispatch.h"
+#include "common/message/dispatch/handle.h"
+#include "common/server/argument.h"
+#include "common/server/handle/call.h"
+#include "queue/api/queue.h"
+
+namespace casual
+{
+   namespace queue::example
+   {
+      namespace local
+      {
+         namespace
+         {
+            struct
+            {
+               std::string queue;
+            } settings;
+
+            namespace service
+            {
+               auto enqueue()
+               {
+                  return []( common::service::invoke::Parameter&& parameter)
+                  {
+                     queue::Message message;
+                     message.payload.data = std::move(parameter.payload.data);
+                     message.payload.type = parameter.payload.type;
+                     queue::enqueue(settings.queue, message);
+                     return common::service::invoke::Result{ nullptr };
+                  };
+               }
+
+               auto dequeue()
+               {
+                  return []( common::service::invoke::Parameter&& parameter)
+                  {
+                     if (auto message = queue::dequeue( settings.queue); ! message.empty())
+                     {
+                        common::buffer::Payload payload { message.front().payload.type };
+                        payload.data = std::move( message.front().payload.data);
+                        return common::service::invoke::Result{ std::move(payload) };
+                     }
+                     return common::service::invoke::Result{ nullptr };
+                  };
+               }
+            }
+
+            common::server::Arguments services()
+            {
+               return {{
+                  {
+                     "casual/example/enqueue",
+                     service::enqueue(),
+                     common::service::transaction::Type::automatic,
+                     common::service::visibility::Type::discoverable,
+                     "example"
+                  },
+                  {
+                     "casual/example/dequeue",
+                     service::dequeue(),
+                     common::service::transaction::Type::automatic,
+                     common::service::visibility::Type::discoverable,
+                     "example"
+                  }
+               }};
+            }
+
+            void main( int argc, char **argv)
+            {
+               common::argument::Parse{
+                  R"(Example server)",
+                  common::argument::Option{ std::tie( settings.queue), {"--queue"}, "queue that casual/example/{enqueue,dequeue} should use"},
+               }(argc, argv);
```

Comment:
I was thinking that one might want to have different queues for enqueue and dequeue, but the real "need" is probably to supply in the request what queue one wants to address.
We'll keep it simple for now :)

Reply:
yes I think we could do another take on that in the future, we might add a blocking dequeue, and possibility to set message attributes as well